drivers/input/input_gpio_qdec.c

```diff
@@ -0,0 +1,225 @@
+{
+	const struct device *dev = k_timer_user_data_get(timer);
+	const struct gpio_qdec_config *cfg = dev->config;
+	struct gpio_qdec_data *data = dev->data;
+	int8_t delta = 0;
+	unsigned int key;
+	uint8_t step = 0x00;
+
+	if (gpio_pin_get_dt(&cfg->gpio[0])) {
+		step |= 0x01;
+	}
+	if (gpio_pin_get_dt(&cfg->gpio[1])) {
+		step |= 0x02;
+	}
+
+	if (data->prev_step == step)
+		return;
+
+	switch ((data->prev_step << 4) | step) {
+	case QDEC_LL_LH:
+	case QDEC_LH_HH:
+	case QDEC_HH_HL:
+	case QDEC_HL_LL:
+		delta = 1;
+		break;
+	case QDEC_LL_HL:
+	case QDEC_LH_LL:
+	case QDEC_HH_LH:
+	case QDEC_HL_HH:
+		delta = -1;
+		break;
+	default:
+		LOG_WRN("%s: lost steps", dev->name);
+	}
+
+	data->prev_step = step;
+
+	key = irq_lock();
+	data->acc += delta;
+	irq_unlock(key);
+
+	if (abs(data->acc) >= cfg->steps_per_period) {
+		k_work_submit(&data->event_work);
+	}
+
+	k_work_reschedule(&data->idle_work, K_MSEC(cfg->idle_timeout_ms));
+}
+
+static void gpio_qdec_event_worker(struct k_work *work)
+{
+	struct gpio_qdec_data *data = CONTAINER_OF(
+			work, struct gpio_qdec_data, event_work);
+	const struct device *dev = data->dev;
+	const struct gpio_qdec_config *cfg = dev->config;
+	unsigned int key;
+	int32_t acc;
+
+	key = irq_lock();
+	acc = data->acc / cfg->steps_per_period;
+	data->acc -= acc * cfg->steps_per_period;
+	irq_unlock(key);
+
+	if (acc != 0) {
+		input_report_rel(data->dev, cfg->axis, acc, true, K_FOREVER);
+	}
+}
+
+static void gpio_qdec_irq_setup(const struct device *dev, bool enable)
+{
+	const struct gpio_qdec_config *cfg = dev->config;
+	unsigned int flags = enable ? GPIO_INT_EDGE_BOTH : GPIO_INT_DISABLE;
+	int ret;
+
+	for (int i = 0; i < 2; i++) {
+		const struct gpio_dt_spec *gpio = &cfg->gpio[i];
+
+		ret = gpio_pin_interrupt_configure_dt(gpio, flags);
+		if (ret != 0) {
+			LOG_ERR("Pin %d interrupt configuration failed: %d", i, ret);
+			return;
+		}
+	}
+}
+
+static void gpio_qdec_idle_worker(struct k_work *work)
+{
+	struct gpio_qdec_data *data = CONTAINER_OF(
+			work, struct gpio_qdec_data, idle_work);
+	const struct device *dev = data->dev;
+
+	k_timer_stop(&data->sample_timer);
+
+	gpio_qdec_irq_setup(dev, true);
+
+	LOG_DBG("polling stop");
+}
+
+static void gpio_qdec_cb(const struct device *gpio_dev, struct gpio_callback *cb,
+			 uint32_t pins)
+{
+	struct gpio_qdec_data *data = CONTAINER_OF(
+			cb, struct gpio_qdec_data, gpio_cb);
+	const struct device *dev = data->dev;
+	const struct gpio_qdec_config *cfg = dev->config;
+
+	gpio_qdec_irq_setup(dev, false);
+
+	k_timer_start(&data->sample_timer, K_NO_WAIT,
+		      K_USEC(cfg->sample_time_us));
+
+	LOG_DBG("polling start");
+}
+
+static int gpio_qdec_init(const struct device *dev)
+{
+	const struct gpio_qdec_config *cfg = dev->config;
+	struct gpio_qdec_data *data = dev->data;
+	int ret;
+
+	data->dev = dev;
+
+	gpio_pin_configure_dt(&cfg->gpio[0], GPIO_INPUT);
+	gpio_pin_configure_dt(&cfg->gpio[1], GPIO_INPUT);
```

Comment:
error handling missing

Reply:
Actually there's already a (proper) `gpio_pin_configure_dt` in the cycle below so this is redundant, removed.